File: Estruturais/Bridge/Python/Bridge.py
class UI:
    RESET = "\033[0m"
    BOLD = "\033[1m"
    PINK = "\033[38;5;213m"
    PURPLE = "\033[38;5;99m"
    WHITE = "\033[97m"
    GREEN = "\033[38;5;84m"
    YELLOW = "\033[38;5;228m"
    RED = "\033[38;5;203m"
    CYAN = "\033[38;5;117m"


class Borda:
    def tipo_borda(self):
        pass


class BordaTradicional(Borda):
    def tipo_borda(self):
        return "Tradicional"


class BordaCatupiry(Borda):
    def tipo_borda(self):
        return "Catupiry"


class SemBorda(Borda):
    def tipo_borda(self):
        return "Sem borda"


class Pizza:
    def __init__(self, borda, sabor):
        self.borda = borda
        self.sabor = sabor

    def get_tamanho(self):
        pass

    def mostrar_pizza(self):

        print(UI.PURPLE + "\n╔════════════════════════════════════════════════════╗" + UI.RESET)
        print(UI.PURPLE + "║               🍕 PIZZA MONTADA 🍕                ║" + UI.RESET)
        print(UI.PURPLE + "╠══════════════╦════════════════╦════════════════════╣" + UI.RESET)

        print(
            UI.PURPLE + "║ " +
            UI.YELLOW + "Tamanho".ljust(12) +
            UI.PURPLE + " ║ " +
            UI.YELLOW + "Borda".ljust(14) +
            UI.PURPLE + " ║ " +
            UI.YELLOW + "Sabor".ljust(18) +
            UI.RESET
        )

        print(UI.PURPLE + "╠══════════════╬════════════════╬════════════════════╣" + UI.RESET)

        print(
            UI.PURPLE + "║ " +
            UI.WHITE + self.get_tamanho().ljust(12) +
            UI.PURPLE + " ║ " +
            UI.WHITE + self.borda.tipo_borda().ljust(14) +
            UI.PURPLE + " ║ " +
            UI.WHITE + self.sabor.ljust(18) +
            UI.RESET
        )

        print(UI.PURPLE + "╚══════════════╩════════════════╩════════════════════╝" + UI.RESET)


class PizzaBrotinho(Pizza):
    def get_tamanho(self):
        return "Brotinho"


class PizzaTradicional(Pizza):
    def get_tamanho(self):
        return "Tradicional"


class PizzaGrande(Pizza):
    def get_tamanho(self):
        return "Grande"


def linha():
    print(UI.PURPLE + "══════════════════════════════════════════════════════" + UI.RESET)


def erro():
    print(UI.RED + "\n❌ Opção inválida! Escolha uma opção do menu.\n" + UI.RESET)


def escolher_tamanho():

    print(UI.PINK + UI.BOLD + "\n🍕 ESCOLHA O TAMANHO DA PIZZA\n" + UI.RESET)
    print(UI.WHITE + "[1]" + UI.RESET + " Brotinho")
    print(UI.WHITE + "[2]" + UI.RESET + " Tradicional")
    print(UI.WHITE + "[3]" + UI.RESET + " Grande")

    op = input(UI.CYAN + "\n➜ Digite sua opção: " + UI.RESET)

    if op == "1":
        return PizzaBrotinho
    elif op == "2":
        return PizzaTradicional
    elif op == "3":
        return PizzaGrande
    else:
        erro()
        return escolher_tamanho()


def escolher_borda():

    linha()

    print(UI.PINK + UI.BOLD + "\n🧀 ESCOLHA A BORDA\n" + UI.RESET)
    print(UI.WHITE + "[1]" + UI.RESET + " Tradicional")
    print(UI.WHITE + "[2]" + UI.RESET + " Catupiry")
    print(UI.WHITE + "[3]" + UI.RESET + " Sem borda")

    op = input(UI.CYAN + "\n➜ Digite sua opção: " + UI.RESET)

    if op == "1":
        return BordaTradicional()
    elif op == "2":
        return BordaCatupiry()
    elif op == "3":
        return SemBorda()
    else:
        erro()
        return escolher_borda()


def escolher_sabor(tipo_pizza, borda):

    linha()

    print(UI.PINK + UI.BOLD + "\n🍕 ESCOLHA O SABOR\n" + UI.RESET)
    print(UI.WHITE + "[1]" + UI.RESET + " Calabresa")
    print(UI.WHITE + "[2]" + UI.RESET + " Frango")
    print(UI.WHITE + "[3]" + UI.RESET + " Portuguesa")

    op = input(UI.CYAN + "\n➜ Digite sua opção: " + UI.RESET)

    if op == "1":
        sabor = "Calabresa"
    elif op == "2":
        sabor = "Frango"
    elif op == "3":
        sabor = "Portuguesa"
    else:
        erro()
        return escolher_sabor(tipo_pizza, borda)

    pizza = tipo_pizza(borda, sabor)
    pizza.mostrar_pizza()

    print(UI.GREEN + "\n✅ Pizza montada utilizando o padrão Bridge!\n" + UI.RESET)


print(UI.PURPLE + UI.BOLD +
      "\n══════════════════════════════════════════════════════" +
      "\n              🍕 PIZZARIA BRIDGE 🍕                  " +
      "\n══════════════════════════════════════════════════════" +
      UI.RESET)

tipo_pizza = escolher_tamanho()
borda = escolher_borda()
escolher_sabor(tipo_pizza, borda)

File: Estruturais/Bridge/JavaScript/Bridge.js
const readline = require("readline");

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const ui = {
    reset: "\x1b[0m",
    bold: "\x1b[1m",
    pink: "\x1b[38;5;213m",
    purple: "\x1b[38;5;99m",
    white: "\x1b[97m",
    green: "\x1b[38;5;84m",
    yellow: "\x1b[38;5;228m",
    red: "\x1b[38;5;203m",
    cyan: "\x1b[38;5;117m"
};

class Borda {
    tipoBorda() {}
}

class BordaTradicional extends Borda {

    tipoBorda() {
        return "Tradicional";
    }
}

class BordaCatupiry extends Borda {

    tipoBorda() {
        return "Catupiry";
    }
}

class SemBorda extends Borda {

    tipoBorda() {
        return "Sem borda";
    }
}

class Pizza {

    constructor(borda, sabor) {
        this.borda = borda;
        this.sabor = sabor;
    }

    getTamanho() {}

    mostrarPizza() {

        console.log(
            ui.purple +
            "\n╔════════════════════════════════════════════════════╗" +
            ui.reset
        );

        console.log(
            ui.purple +
            "║               🍕 PIZZA MONTADA 🍕                ║" +
            ui.reset
        );

        console.log(
            ui.purple +
            "╠══════════════╦════════════════╦════════════════════╣" +
            ui.reset
        );

        console.log(
            ui.purple + "║ " +
            ui.yellow + "Tamanho".padEnd(12) +
            ui.purple + " ║ " +
            ui.yellow + "Borda".padEnd(14) +
            ui.purple + " ║ " +
            ui.yellow + "Sabor".padEnd(18) +
            ui.purple + " ║" +
            ui.reset
        );

        console.log(
            ui.purple +
            "╠══════════════╬════════════════╬════════════════════╣" +
            ui.reset
        );

        console.log(
            ui.purple + "║ " +
            ui.white + this.getTamanho().padEnd(12) +
            ui.purple + " ║ " +
            ui.white + this.borda.tipoBorda().padEnd(14) +
            ui.purple + " ║ " +
            ui.white + this.sabor.padEnd(18) +
            ui.purple + " ║" +
            ui.reset
        );

        console.log(
            ui.purple +
            "╚══════════════╩════════════════╩════════════════════╝" +
            ui.reset
        );
    }
}

class PizzaBrotinho extends Pizza {

    getTamanho() {
        return "Brotinho";
    }
}

class PizzaTradicional extends Pizza {

    getTamanho() {
        return "Tradicional";
    }
}

class PizzaGrande extends Pizza {

    getTamanho() {
        return "Grande";
    }
}

function linha() {

    console.log(
        ui.purple +
        "══════════════════════════════════════════════════════" +
        ui.reset
    );
}

function erro() {

    console.log(
        ui.red +
        "\n❌ Opção inválida! Escolha uma opção do menu.\n" +
        ui.reset
    );
}

function escolherTamanho() {

    console.log(
        ui.pink +
        ui.bold +
        "\n🍕 ESCOLHA O TAMANHO DA PIZZA\n" +
        ui.reset
    );

    console.log(ui.white + "[1]" + ui.reset + " Brotinho");
    console.log(ui.white + "[2]" + ui.reset + " Tradicional");
    console.log(ui.white + "[3]" + ui.reset + " Grande");

    rl.question(
        ui.cyan + "\n➜ Digite sua opção: " + ui.reset,
        (opTamanho) => {

            let tipoPizza = null;

            if (opTamanho === "1") {
                tipoPizza = PizzaBrotinho;
            }

            else if (opTamanho === "2") {
                tipoPizza = PizzaTradicional;
            }

            else if (opTamanho === "3") {
                tipoPizza = PizzaGrande;
            }

            else {
                erro();
                return escolherTamanho();
            }

            escolherBorda(tipoPizza);
        }
    );
}

function escolherBorda(tipoPizza) {

    linha();

    console.log(
        ui.pink +
        ui.bold +
        "\n🧀 ESCOLHA A BORDA\n" +
        ui.reset
    );

    console.log(ui.white + "[1]" + ui.reset + " Tradicional");
    console.log(ui.white + "[2]" + ui.reset + " Catupiry");
    console.log(ui.white + "[3]" + ui.reset + " Sem borda");

    rl.question(
        ui.cyan + "\n➜ Digite sua opção: " + ui.reset,
        (opBorda) => {

            let borda = null;

            if (opBorda === "1") {
                borda = new BordaTradicional();
            }

            else if (opBorda === "2") {
                borda = new BordaCatupiry();
            }

            else if (opBorda === "3") {
                borda = new SemBorda();
            }

            else {
                erro();
                return escolherBorda(tipoPizza);
            }

            escolherSabor(tipoPizza, borda);
        }
    );
}

function escolherSabor(tipoPizza, borda) {

    linha();

    console.log(
        ui.pink +
        ui.bold +
        "\n🍕 ESCOLHA O SABOR\n" +
        ui.reset
    );

    console.log(ui.white + "[1]" + ui.reset + " Calabresa");
    console.log(ui.white + "[2]" + ui.reset + " Frango");
    console.log(ui.white + "[3]" + ui.reset + " Portuguesa");

    rl.question(
        ui.cyan + "\n➜ Digite sua opção: " + ui.reset,
        (opSabor) => {

            let sabor = "";

            if (opSabor === "1") {
                sabor = "Calabresa";
            }

            else if (opSabor === "2") {
                sabor = "Frango";
            }

            else if (opSabor === "3") {
                sabor = "Portuguesa";
            }

            else {
                erro();
                return escolherSabor(tipoPizza, borda);
            }

            const pizza = new tipoPizza(
                borda,
                sabor
            );

            pizza.mostrarPizza();

            console.log(
                ui.green +
                "\n✅ Pizza montada utilizando o padrão Bridge!\n" +
                ui.reset
            );

            rl.close();
        }
    );
}

console.clear();

console.log(
    ui.purple +
    ui.bold +
    "\n══════════════════════════════════════════════════════" +
    "\n              🍕 PIZZARIA BRIDGE 🍕                  " +
    "\n══════════════════════════════════════════════════════" +
    ui.reset
);

escolherTamanho();